### PR TITLE
[WIP] Make model builder lambda APIs work with public fields on entities

### LINKED ISSUE #11410: model builder lambda APIs don't work with public fields on entities
repro:

```cs
        [Fact]
        public virtual void Field_key_test()
        {
            using (var ctx = new EntityWithFieldKeyContext())
            {
                ctx.Database.EnsureDeleted();
                ctx.Database.EnsureCreated();
            }
        }

        public class EntityWithFieldKey
        {
            public long Id;
        }

        public class EntityWithFieldKeyContext : DbContext
        {
            public DbSet<EntityWithFieldKey> MyEntities { get; set; }

            protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
            {
                optionsBuilder.UseSqlServer(@"Server=.;Database=Repro_field_key;Trusted_Connection=True;MultipleActiveResultSets=True");
            }

            protected override void OnModelCreating(ModelBuilder modelBuilder)
            {
                modelBuilder.Entity<EntityWithFieldKey>().Property(e => e.Id);
                modelBuilder.Entity<EntityWithFieldKey>().HasKey(e => e.Id);
            }
        }
```

throws:

```
System.ArgumentException : The expression 'e => e.Id' is not a valid property expression. The expression should represent a simple property access: 't => t.MyProperty'.
Parameter name: propertyAccessExpression
	at Microsoft.EntityFrameworkCore.Internal.ExpressionExtensions.GetPropertyAccess(LambdaExpression propertyAccessExpression)
	at Microsoft.EntityFrameworkCore.Metadata.Builders.EntityTypeBuilder`1.Property[TProperty](Expression`1 propertyExpression)
	at Microsoft.EntityFrameworkCore.Query.QueryBugsTest.EntityWithFieldKeyContext.OnModelCreating(ModelBuilder modelBuilder)
	at Microsoft.EntityFrameworkCore.Infrastructure.ModelCustomizer.Customize(ModelBuilder modelBuilder, DbContext context)
	at Microsoft.EntityFrameworkCore.Infrastructure.RelationalModelCustomizer.Customize(ModelBuilder modelBuilder, DbContext context)
	at Microsoft.EntityFrameworkCore.Infrastructure.ModelSource.CreateModel(DbContext context, IConventionSetBuilder conventionSetBuilder, IModelValidator validator)
	at Microsoft.EntityFrameworkCore.Infrastructure.ModelSource.<>c__DisplayClass5_0.<GetModel>b__1()
	at System.Lazy`1.CreateValue()
	at System.Lazy`1.LazyInitValue()
	at System.Lazy`1.get_Value()
	at Microsoft.EntityFrameworkCore.Infrastructure.ModelSource.GetModel(DbContext context, IConventionSetBuilder conventionSetBuilder, IModelValidator validator)
	at Microsoft.EntityFrameworkCore.Internal.DbContextServices.CreateModel()
	at Microsoft.EntityFrameworkCore.Internal.DbContextServices.get_Model()
	at Microsoft.EntityFrameworkCore.Infrastructure.EntityFrameworkServicesBuilder.<>c.<TryAddCoreServices>b__7_1(IServiceProvider p)
	at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitFactory(FactoryCallSite factoryCallSite, ServiceProviderEngineScope scope)
	at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(IServiceCallSite callSite, TArgument argument)
	at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScoped(ScopedCallSite scopedCallSite, ServiceProviderEngineScope scope)
	at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(IServiceCallSite callSite, TArgument argument)
	at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, ServiceProviderEngineScope scope)
	at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(IServiceCallSite callSite, TArgument argument)
	at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScoped(ScopedCallSite scopedCallSite, ServiceProviderEngineScope scope)
	at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(IServiceCallSite callSite, TArgument argument)
	at Microsoft.Extensions.DependencyInjection.ServiceLookup.DynamicServiceProviderEngine.<>c__DisplayClass1_0.<RealizeService>b__0(ServiceProviderEngineScope scope)
	at Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngine.GetService(Type serviceType, ServiceProviderEngineScope serviceProviderEngineScope)
	at Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngineScope.GetService(Type serviceType)
	at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService(IServiceProvider provider, Type serviceType)
	at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService[T](IServiceProvider provider)
	at Microsoft.EntityFrameworkCore.DbContext.get_DbContextDependencies()
	at Microsoft.EntityFrameworkCore.DbContext.get_InternalServiceProvider()
	at Microsoft.EntityFrameworkCore.DbContext.Microsoft.EntityFrameworkCore.Infrastructure.IInfrastructure<System.IServiceProvider>.get_Instance()
	at Microsoft.EntityFrameworkCore.Infrastructure.DatabaseFacade.Microsoft.EntityFrameworkCore.Infrastructure.IInfrastructure<System.IServiceProvider>.get_Instance()
	at Microsoft.EntityFrameworkCore.Internal.InternalAccessorExtensions.GetService[TService](IInfrastructure`1 accessor)
	at Microsoft.EntityFrameworkCore.Infrastructure.AccessorExtensions.GetService[TService](IInfrastructure`1 accessor)
	at Microsoft.EntityFrameworkCore.Infrastructure.DatabaseFacade.get_DatabaseCreator()
	at Microsoft.EntityFrameworkCore.Infrastructure.DatabaseFacade.EnsureDeleted()
	at Microsoft.EntityFrameworkCore.Query.QueryBugsTest.Field_key_test()
```

### PR DESCRIPTION
**Pull request by ilmax**
**Please check if the PR fulfills these requirements**

- [x] The code builds and tests pass (verified by our automated build checks)
- [ ] Commit messages follow this format
```
    Summary of the changes
    - Detail 1
    - Detail 2

    Fixes #bugnumber
```
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Code meets the expectations our engineering guidelines. https://github.com/aspnet/Home/wiki/Engineering-guidelines.

Fixes #11410

### CODE REVIEW

**Comment by ilmax** [comment]
Open points:

-  Figure out what model builder methods should accept expression of either properties or fields
-  Figure out what's the best place for the added unit test methods

File: src/EFCore/Extensions/Internal/ExpressionExtensions.cs [hunk 1]
```diff
@@ -100,30 +164,32 @@ var propertyPaths
...
+        private static IReadOnlyList<T> MatchPropertyOrFieldAccessList<T>(
+            this LambdaExpression lambdaExpression, Func<Expression, Expression, T> propertyMatcher)
+            where T : MemberInfo
         {
             Debug.Assert(lambdaExpression.Body != null);
+            Debug.Assert(lambdaExpression.Parameters.Count == 1);
 
             var parameterExpression
-                = lambdaExpression.Parameters.Single();
+                = lambdaExpression.Parameters[0];
```
Comment (ralmsdeveloper) [inline src/EFCore/Extensions/Internal/ExpressionExtensions.cs]: Single line. 😊

**Comment by ilmax** [comment]
@ajcvickers Did you had a chance to take a look at this one?

**Comment by ajcvickers** [comment]
@ilmax We will look again next week, since right now we are not putting new features into 2.1, but we haven't branched yet.

File: src/EFCore/Extensions/Internal/ExpressionExtensions.cs [hunk 2]
```diff
@@ -75,6 +75,70 @@ public static PropertyInfo GetPropertyAccess([NotNull] this LambdaExpression pro
             return propertyInfo;
         }
 
+        /// <summary>
+        ///     This API supports the Entity Framework Core infrastructure and is not intended to be used
+        ///     directly from your code. This API may change or be removed in future releases.
+        /// </summary>
+        public static MemberInfo GetPropertyOrFieldAccess([NotNull] this LambdaExpression lambdaExpression)
```
Comment (smitpatel) [inline src/EFCore/Extensions/Internal/ExpressionExtensions.cs]: Why do we need code duplication instead of one path for both property/field

File: test/EFCore.Tests/ModelBuilding/FieldTestBase.cs [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+using Microsoft.EntityFrameworkCore.TestUtilities;
+using Xunit;
+
+// ReSharper disable InconsistentNaming
+namespace Microsoft.EntityFrameworkCore.ModelBuilding
+{
+    public abstract partial class ModelBuilderTest
+    {
+        // TODO is this the correct place for these tests?
+        public abstract class FieldTestBase : ModelBuilderTestBase
```
Comment (smitpatel) [inline test/EFCore.Tests/ModelBuilding/FieldTestBase.cs]: This should go inside nonrelationshiptest base if no relations. And in relationship testbases according to relationship type

**Comment by smitpatel** [comment]
> Figure out what model builder methods should accept expression of either properties or fields

Will let you know in a day or two after discussing with team.

**Comment by ilmax** [comment]
@smitpatel, I would like to make some progress with this one. What do you suggest to avoid duplicating `GetPropertyAccess` and also we should figure out all the *builder's methods that needs to work with both property and field based expressions

**Comment by smitpatel** [comment]
@ilmax - We should enable field based expression for all the modelbuilder APIs which works with string & property based expressions. That would also help us uniquifying pipeline.
cc: @divega

**Comment by ilmax** [comment]
@smitpatel I'm looking into this and I've updated almost all the methods that takes either a string or an expression. Two places cannot be updated without introducing breaking changes, they are `HasOne`, `HasMany` because `ReferenceNavigationBuilder` and `CollectionNavigationBuilder` both exposes a `PropertyInfo` object.

Should I update those types as well?

**Comment by AndriySvyryd** [comment]
@smitpatel, @ilmax While those breaking changes should be ok there are more things to consider for the navigations API, see https://github.com/aspnet/EntityFrameworkCore/issues/6674. Therefore it would be better to do in a separate PR

**Comment by smitpatel** [comment]
@ilmax - Let's exclude navigation APIs from using Fields and do it for all other APIs

**Comment by ilmax** [comment]
Great! I will work on this one in the coming days!

**Comment by ilmax** [comment]
@smitpatel, @ajcvickers, @AndriySvyryd this needs a bit (read waaaaaaaaaaaayyyyyyyyyy more) test coverage, but I would like to know what to do with the duplicate methods, can we live with that for now? what do you think?

**Comment by smitpatel** [comment]
@ilmax If it is difficult to remove duplicated code then we can certainly live with it and do clean up later.

I see you have added some breaking changes. Perhaps that requires review if we want to take breaking changes or we should delay them till 3.0 release.

@AndriySvyryd - Any idea of testing? ModelBuilderTest is the right location but can we do some magic to run all things for fields/properties?

**Comment by ilmax** [comment]
@smitpatel I will try to remove duplication when I'm done with tests, I think that with some generics gymnastics we can probably unify the methods. @AndriySvyryd thank you for the suggestion, I will proceed as you suggested!

**Comment by ilmax** [comment]
Hi @smitpatel, I had a couple of issues recently and had no chance to work on this, yesterday I made some changes to avoid code duplication and I added few more tests.
I also read your [comment](https://github.com/aspnet/EntityFrameworkCore/pull/11610#issuecomment-389301592) and realised I should undo some of the changes, so here it is the list of methods that now supports handling expression with either properties or fields:

```
ReferenceNavigationBuilder<TEntity, TRelatedEntity>.WithMany

CollectionNavigationBuilder<TEntity, TRelatedEntity>.WithOne

EntityEntry<TEntity>.Property
EntityEntry<TEntity>.Reference
EntityEntry<TEntity>.Collection

EntityTypeBuilder<TEntity>.Property
EntityTypeBuilder<TEntity>.Ignore
EntityTypeBuilder<TEntity>.OwnsOne
EntityTypeBuilder<TEntity>.HasKey
EntityTypeBuilder<TEntity>.HasAlternateKey
EntityTypeBuilder<TEntity>.HasIndex

QueryTypeBuilder<TQuery>.Property
QueryTypeBuilder<TQuery>.Ignore

ReferenceNavigationBuilder<TEntity, TRelatedEntity>.WithMany
ReferenceNavigationBuilder<TEntity, TRelatedEntity>.WithOne

ReferenceOwnershipBuilder<TEntity, TRelatedEntity>.Property
ReferenceOwnershipBuilder<TEntity, TRelatedEntity>.Ignore
ReferenceOwnershipBuilder<TEntity, TRelatedEntity>.OwnsOne
ReferenceOwnershipBuilder<TEntity, TRelatedEntity>.HasForeignKey
ReferenceOwnershipBuilder<TEntity, TRelatedEntity>.HasPrincipalKey
ReferenceOwnershipBuilder<TEntity, TRelatedEntity>.HasIndex

ReferenceReferenceBuilder<TEntity, TRelatedEntity>.HasForeignKey
ReferenceReferenceBuilder<TEntity, TRelatedEntity>.HasPrincipalKey

ReferenceCollectionBuilder<TPrincipalEntity, TDependentEntity>.HasForeignKey
ReferenceCollectionBuilder<TPrincipalEntity, TDependentEntity>.HasPrincipalKey

RelationalEntityTypeBuilderExtensions.HasDiscriminator

RelationalQueryTypeBuilderExtensions.HasDiscriminator

```

I would like to agree on which one to undo, there are some method like `HasForeignKey` very close to the navigation property api that we might still want to allow accepting lambda with fields because we allow `HasKey` to be called with a field based expression.

Moreover I'm hitting a couple of issues in the CI around the MonsterFixupChangedChangingSqlServerTest and not sure why, and getting unit test run locally on VS is getting harder and harder, I will investigate what's happening here but any help would be highly appreciated :)

**Comment by smitpatel** [comment]
@ilmax - Looking at test failures it seems like perhaps in `WithMany` API with lambda is not getting correct type inferred. That is causing type difference. You can compare the models with & without change to figure out why the type is different now.

**Comment by ilmax** [comment]
@smitpatel Thank you for pointing that out, I'm having hard times to have unit testing work in visual studio but I'm on it. What do you think about the changes in `ExpressionExtensions`?

File: src/EFCore/Metadata/Builders/CollectionNavigationBuilder.cs [hunk 4]
```diff
@@ -130,8 +130,8 @@ protected virtual InternalRelationshipBuilder WithOneBuilder([CanBeNull] string
         ///     This API supports the Entity Framework Core infrastructure and is not intended to be used
         ///     directly from your code. This API may change or be removed in future releases.
         /// </summary>
-        protected virtual InternalRelationshipBuilder WithOneBuilder([CanBeNull] PropertyInfo navigationProperty)
-            => WithOneBuilder(PropertyIdentity.Create(navigationProperty));
+        protected virtual InternalRelationshipBuilder WithOneBuilder([CanBeNull] MemberInfo navigationMemberInfo)
```
Comment (smitpatel) [inline src/EFCore/Metadata/Builders/CollectionNavigationBuilder.cs]: @andriysvyryd - Is this acceptable break?

File: src/EFCore/Metadata/Builders/ReferenceCollectionBuilder.cs [hunk 5]
```diff
@@ -165,8 +165,8 @@ protected virtual InternalRelationshipBuilder HasForeignKeyBuilder([NotNull] IRe
         ///     This API supports the Entity Framework Core infrastructure and is not intended to be used
         ///     directly from your code. This API may change or be removed in future releases.
         /// </summary>
-        protected virtual InternalRelationshipBuilder HasForeignKeyBuilder([NotNull] IReadOnlyList<PropertyInfo> foreignKeyProperties)
-            => Builder.HasForeignKey(foreignKeyProperties, _dependentEntityType, ConfigurationSource.Explicit);
+        protected virtual InternalRelationshipBuilder HasForeignKeyBuilder([NotNull] IReadOnlyList<MemberInfo> foreignKeyMembers)
```
Comment (smitpatel) [inline src/EFCore/Metadata/Builders/ReferenceCollectionBuilder.cs]: @andriysvyryd - And this variation of break?

File: test/EFCore.Tests/ModelBuilding/NonRelationshipTestBase.cs [hunk 6]
```diff
@@ -1208,6 +1209,99 @@ public virtual void Can_ignore_explicit_interface_implementation()
...
+            public virtual void Entity_field_expression_ignore_test()
+            {
+                var modelBuilder = InMemoryTestHelpers.Instance.CreateConventionBuilder();
+                modelBuilder.Entity<EntityWithFieldKey>().Ignore(e => e.Year);
+
+                Assert.Empty(modelBuilder.Model.FindEntityType(typeof(EntityWithFieldKey)).GetProperties());
+            }
+
+            [Fact]
+            public virtual void Query_field_expression_property_test()
```
Comment (smitpatel) [inline test/EFCore.Tests/ModelBuilding/NonRelationshipTestBase.cs]: Also add variation for composite case (for relevant places)

**Comment by smitpatel** [comment]
@ilmax - Changes in ExpressionExtensions look good. Perhaps more unification can be done but I may be missing something. I will investigate during merge so you are good on that side.

Let see what @AndriySvyryd says for breaking changes for navigation API & relationship APIs like HasFK/HasPK. If they are fine to change then we would need tests for relationship based too.

File: src/EFCore/Metadata/Builders/CollectionNavigationBuilder.cs (hunk 4 above)
Comment (AndriySvyryd) [reply src/EFCore/Metadata/Builders/CollectionNavigationBuilder.cs]: Yes

File: src/EFCore/Metadata/Builders/ReferenceCollectionBuilder.cs (hunk 5 above)
Comment (AndriySvyryd) [reply src/EFCore/Metadata/Builders/ReferenceCollectionBuilder.cs]: Yes

**Comment by ilmax** [comment]
@smitpatel Good to know, I'm tracking down the test failure and it's caused by an incorrect collection type instantiated for the nav property, it's a `HashSet` intead of an `ObservableCollection` because the `CollectionTypeFactory` get's an interface instead of a concrete type, so this check

```
if (typeof(INotifyPropertyChanged).GetTypeInfo().IsAssignableFrom(entityType.GetTypeInfo()))
{
	var observableHashSetOfT = typeof(ObservableHashSet<>).MakeGenericType(elementType);
	if (collectionType.GetTypeInfo().IsAssignableFrom(observableHashSetOfT.GetTypeInfo()))
	{
		return observableHashSetOfT;
	}
}
```
is failing and thus EF creates an HashSet.
Now I have to track down why it get's an interface instead of a concrete type

**Comment by ilmax** [comment]
@AndriySvyryd any hint on this one? Are we ok with the breaking changes introduced here?

File: src/EFCore/Extensions/Internal/ExpressionExtensions.cs [hunk 7]
```diff
@@ -100,42 +117,66 @@ var propertyPaths
...
-            return propertyPath != null ? new[] { propertyPath } : null;
+            return memberPath != null ? new[] { memberPath } : null;
         }
 
         /// <summary>
         ///     This API supports the Entity Framework Core infrastructure and is not intended to be used
         ///     directly from your code. This API may change or be removed in future releases.
         /// </summary>
-        private static PropertyInfo MatchSimplePropertyAccess(
+        private static TMemberInfo MatchSimplePropertyAccess<TMemberInfo>(
```
Comment (AndriySvyryd) [inline src/EFCore/Extensions/Internal/ExpressionExtensions.cs]: Rename to `MatchSimpleMemberAccess`

File: src/EFCore/Extensions/Internal/ExpressionExtensions.cs [hunk 8]
```diff
@@ -169,30 +210,31 @@ var propertyPath
...
                     .Parameters
                     .Single()
-                    .MatchPropertyAccess(propertyAccessExpression.Body);
+                    .MatchPropertyAccess<PropertyInfo>(propertyAccessExpression.Body);
 
             return propertyPath != null;
         }
 
-        private static IReadOnlyList<PropertyInfo> MatchPropertyAccess(
+        private static IReadOnlyList<TMemberInfo> MatchPropertyAccess<TMemberInfo>(
```
Comment (AndriySvyryd) [inline src/EFCore/Extensions/Internal/ExpressionExtensions.cs]: Rename to `MatchMemberAccess`

File: test/EFCore.Tests/ModelBuilding/NonRelationshipTestBase.cs [hunk 9]
```diff
@@ -1208,6 +1209,159 @@ public virtual void Can_ignore_explicit_interface_implementation()
...
+                Assert.NotNull(first.FieldInfo);
+                Assert.Equal(nameof(EntityWithFieldKey.CompanyId), second.Name);
+                Assert.Null(second.PropertyInfo);
+                Assert.NotNull(second.FieldInfo);
+            }
+
+            [Fact]
+            public virtual void Entity_field_expression_owns_one_test()
+            {
+                // TODO this should be moved in the relationship test base
```
Comment (AndriySvyryd) [inline test/EFCore.Tests/ModelBuilding/NonRelationshipTestBase.cs]: Also add tests for `HasOne`/`HasMany`, `WithOne`/`WithMany` and tests calling all the methods on `ReferenceOwnershipBuilder` returned by `OwnsOne`

**Comment by ajcvickers** [comment]
Triage: closing this for now since it has been inactive for a while. We would be happy to re-open as needed.
